File: apps/chat/src/collectors.test.ts
import {
  ChatSourceIdSchema,
  type ChatSourceId,
  type ChatStreamEvent,
} from "@coldbrew/packages/chat.js";
import { erro } from "@lebedevna/neverthrow-utils";
import { ok } from "neverthrow";
import { describe, expect, it, vi } from "vitest";

import type { ChatEventBroker } from "./chat-application.js";
import {
  runChatCollectors,
  type ChatCollectorLeases,
  type ChatCollectorRefreshControl,
  type ChatCollectorStore,
} from "./collectors.js";
import type { ChatProviderAdapter, ConnectedChatSource } from "./provider.js";

const sourceId = ChatSourceIdSchema.parse("00000000-0000-4000-8000-000000000001");
const connectedSource: ConnectedChatSource = {
  source: {
    sourceId,
    connectionId: "10000000-0000-4000-8000-000000000001",
    provider: "youtube",
    providerSourceId: "channel-1",
    displayName: "Channel",
    sourceUrl: "https://www.youtube.com/channel/channel-1",
    position: 0,
    enabled: true,
  },
  capabilities: ["read"],
  credentials: { accessToken: "token", scopes: [], tokenVersion: 1 },
};

const store: ChatCollectorStore = {
  async getAllEnabledSources() {
    return [{ userId: 42, connectedSource }];
  },
};

const leases: ChatCollectorLeases = {
  async acquire() {
    return {
      async maintain(signal) {
        if (signal.aborted) {
          return;
        }
        await new Promise<void>((resolve) =>
          signal.addEventListener("abort", () => resolve(), { once: true }),
        );
      },
      async release() {},
    };
  },
};

function idleRefreshControl(): ChatCollectorRefreshControl {
  return {
    async *refreshes(signal) {
      if (signal.aborted) {
        return;
      }
      await new Promise<void>((resolve) =>
        signal.addEventListener("abort", () => resolve(), { once: true }),
      );
    },
  };
}

describe("chat collectors", () => {
  it("publishes stream errors and continues consuming a retrying provider", async () => {
    const controller = new AbortController();
    const published: ChatStreamEvent[] = [];
    const broker: ChatEventBroker = {
      async publish(_userId, event) {
        published.push(event);
        if (event.type === "state" && event.state === "offline") {
          controller.abort();
        }
      },
      async *stream() {},
    };
    const provider: ChatProviderAdapter = {
      provider: "youtube",
      collection: "pull",
      async *stream() {
        yield erro({ type: "provider unavailable" as const, detail: "temporary" });
        yield ok({ type: "state" as const, sourceId, state: "offline" as const });
      },
      async sendMessage() {
        return ok(undefined);
      },
      async moderate() {
        return ok({});
      },
    };

    await runChatCollectors(
      store,
      broker,
      leases,
      idleRefreshControl(),
      [provider],
      controller.signal,
    );

    expect(published).toEqual([
      { type: "state", sourceId, state: "error", detail: "temporary" },
      { type: "state", sourceId, state: "offline" },
    ]);
  });

  it("restarts the lease owner when a manual refresh is requested", async () => {
    const controller = new AbortController();
    let requestRefresh: ((value: ChatSourceId) => void) | undefined;
    const refreshControl: ChatCollectorRefreshControl = {
      async *refreshes(signal) {
        const refreshedSourceId = await new Promise<ChatSourceId>((resolve) => {
          requestRefresh = resolve;
        });
        if (!signal.aborted) {
          yield refreshedSourceId;
        }
      },
    };
    const broker: ChatEventBroker = {
      async publish(_userId, event) {
        if (event.type !== "state") {
          return;
        }
        if (event.state === "offline") {
          requestRefresh?.(sourceId);
        }
        if (event.state === "live") {
          controller.abort();
        }
      },
      async *stream() {},
    };
    const starts = vi.fn();
    const provider: ChatProviderAdapter = {
      provider: "youtube",
      collection: "pull",
      stream(_source, signal) {
        const start = starts.mock.calls.length;
        starts();
        return (async function* () {
          if (start === 0) {
            yield ok({ type: "state" as const, sourceId, state: "offline" as const });
            if (!signal.aborted) {
              await new Promise<void>((resolve) =>
                signal.addEventListener("abort", () => resolve(), { once: true }),
              );
            }
            return;
          }
          yield ok({ type: "state" as const, sourceId, state: "live" as const });
        })();
      },
      async sendMessage() {
        return ok(undefined);
      },
      async moderate() {
        return ok({});
      },
    };

    await runChatCollectors(store, broker, leases, refreshControl, [provider], controller.signal);

    expect(starts).toHaveBeenCalledTimes(2);
  });
});

File: internal/chat/collectors_test.go
package chat

import (
	"context"
	"reflect"
	"sync"
	"testing"
)

type collectorTestStore struct{ source OwnedConnectedSource }

func (store *collectorTestStore) GetAllEnabledSources(context.Context) ([]OwnedConnectedSource, error) {
	return []OwnedConnectedSource{store.source}, nil
}

type collectorTestLease struct{}

func (*collectorTestLease) Maintain(ctx context.Context) error { <-ctx.Done(); return nil }
func (*collectorTestLease) Release() error                     { return nil }

type collectorTestLeases struct{}

func (*collectorTestLeases) Acquire(context.Context, string, string) (CollectorLease, error) {
	return &collectorTestLease{}, nil
}

type collectorTestControl struct{ refreshes <-chan string }

func (control *collectorTestControl) Refreshes(context.Context) (<-chan string, error) {
	return control.refreshes, nil
}

type collectorTestBroker struct {
	mu      sync.Mutex
	events  []StreamEvent
	onEvent func(StreamEvent)
}

func (broker *collectorTestBroker) Publish(_ context.Context, _ int, event StreamEvent, _ string) error {
	broker.mu.Lock()
	broker.events = append(broker.events, event)
	broker.mu.Unlock()
	if broker.onEvent != nil {
		broker.onEvent(event)
	}
	return nil
}
func (*collectorTestBroker) Stream(context.Context, int) <-chan StreamEvent {
	output := make(chan StreamEvent)
	close(output)
	return output
}

type collectorTestProvider struct {
	stream func(context.Context) (<-chan StreamEvent, <-chan error)
}

func (*collectorTestProvider) Name() string       { return "youtube" }
func (*collectorTestProvider) Collection() string { return "pull" }
func (provider *collectorTestProvider) Stream(ctx context.Context, _ ConnectedSource) (<-chan StreamEvent, <-chan error) {
	return provider.stream(ctx)
}
func (*collectorTestProvider) SendMessage(context.Context, ConnectedSource, string) error { return nil }
func (*collectorTestProvider) Moderate(context.Context, ConnectedSource, ModerationCommand, string) (ProviderCommandSuccess, error) {
	return ProviderCommandSuccess{}, nil
}

func testOwnedSource() OwnedConnectedSource {
	return OwnedConnectedSource{UserID: 42, ConnectedSource: ConnectedSource{Source: Source{SourceID: youtubeSourceID, ConnectionID: "10000000-0000-4000-8000-000000000001", Provider: "youtube", ProviderSourceID: "channel-1", DisplayName: "Channel", SourceURL: "https://www.youtube.com/channel/channel-1", Enabled: true}, Capabilities: []Capability{CapabilityRead}, Credentials: ProviderCredentials{AccessToken: "token", TokenVersion: 1}}}
}

func TestCollectorsPublishErrorsAndContinue(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	refreshes := make(chan string)
	broker := &collectorTestBroker{}
	broker.onEvent = func(event StreamEvent) {
		if event.Type == "state" && event.State == "offline" {
			cancel()
		}
	}
	provider := &collectorTestProvider{stream: func(ctx context.Context) (<-chan StreamEvent, <-chan error) {
		events := make(chan StreamEvent)
		errorsChannel := make(chan error)
		go func() {
			defer close(events)
			defer close(errorsChannel)
			select {
			case errorsChannel <- &ProviderError{Type: "provider unavailable", Detail: "temporary"}:
			case <-ctx.Done():
				return
			}
			select {
			case events <- StreamEvent{Type: "state", SourceID: youtubeSourceID, State: "offline"}:
			case <-ctx.Done():
				return
			}
		}()
		return events, errorsChannel
	}}
	err := RunCollectors(ctx, &collectorTestStore{source: testOwnedSource()}, broker, &collectorTestLeases{}, &collectorTestControl{refreshes: refreshes}, []Provider{provider})
	if err != nil {
		t.Fatal(err)
	}
	expected := []StreamEvent{{Type: "state", SourceID: youtubeSourceID, State: "error", Detail: "temporary"}, {Type: "state", SourceID: youtubeSourceID, State: "offline"}}
	if !reflect.DeepEqual(broker.events, expected) {
		t.Fatalf("events = %#v; want %#v", broker.events, expected)
	}
}

func TestCollectorsRestartOnManualRefresh(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	refreshes := make(chan string, 1)
	broker := &collectorTestBroker{}
	starts := 0
	var startsMutex sync.Mutex
	provider := &collectorTestProvider{stream: func(ctx context.Context) (<-chan StreamEvent, <-chan error) {
		events := make(chan StreamEvent)
		errorsChannel := make(chan error)
		startsMutex.Lock()
		start := starts
		starts++
		startsMutex.Unlock()
		go func() {
			defer close(events)
			defer close(errorsChannel)
			state := "offline"
			if start > 0 {
				state = "live"
			}
			select {
			case events <- StreamEvent{Type: "state", SourceID: youtubeSourceID, State: state}:
			case <-ctx.Done():
				return
			}
			if start == 0 {
				<-ctx.Done()
			}
		}()
		return events, errorsChannel
	}}
	broker.onEvent = func(event StreamEvent) {
		if event.State == "offline" {
			refreshes <- youtubeSourceID
		}
		if event.State == "live" {
			cancel()
		}
	}
	if err := RunCollectors(ctx, &collectorTestStore{source: testOwnedSource()}, broker, &collectorTestLeases{}, &collectorTestControl{refreshes: refreshes}, []Provider{provider}); err != nil {
		t.Fatal(err)
	}
	startsMutex.Lock()
	defer startsMutex.Unlock()
	if starts != 2 {
		t.Fatalf("provider started %d times; want 2", starts)
	}
}
